bpo-43950: Initial base implementation for PEP 657

pablogsal opened the pull request:
⚠️  This is not the full feature and contains no optimizations for the time being, just the base implementation to avoid merge conflicts in the future ⚠️ 

**All the internal structures, and data representation are subjected to change in future PRs as we add optimizations**

This PR is part of PEP 657 and augments the compiler to emit ending
line numbers as well as starting and ending columns from the AST
into compiled code objects. This allows bytecodes to be correlated
to the exact source code ranges that generated them.

This information is made available through the following public APIs:

* The `co_positions` method on code objects.
* The C API function `PyCode_Addr2Location`.

Co-authored-by: Pablo Galindo <Pablogsal@gmail.com>
Co-authored-by: Batuhan Taskaya <isidentical@gmail.com>
Co-authored-by: Ammar Asker <ammar@ammaraskar.com>


<!-- issue-number: [bpo-43950](https://bugs.python.org/issue43950) -->
https://bugs.python.org/issue43950
<!-- /issue-number -->

ammaraskar commented on Objects/typeobject.c:
```diff
@@ -1065,12 +1065,6 @@ static PyGetSetDef type_getsets[] = {
 static PyObject *
 type_repr(PyTypeObject *type)
 {
-    if (type->tp_name == NULL) {
```
I think this accidentally got caught up in the merge

ammaraskar commented on Python/traceback.c:
```diff
@@ -926,4 +1008,3 @@ _Py_DumpTracebackThreads(int fd, PyInterpreterState *interp,
 
     return NULL;
 }
-
```
Aah applying the github suggestion wiped this blank line.

isidentical replied:
Probably my bad, it was a rough merge with all the stuff changed. I think you should have commit privs on colnotab, so feel free to fix these irrelevant stuff in commit

ammaraskar committed:
bpo-43950: Add code.co_positions (PEP 657)

This PR is part of PEP 657 and augments the compiler to emit ending
line numbers as well as starting and ending columns from the AST
into compiled code objects. This allows bytecodes to be correlated
to the exact source code ranges that generated them.

This information is made available through the following public APIs:

* The `co_positions` method on code objects.
* The C API function `PyCode_Addr2Location`.

Co-authored-by: Batuhan Taskaya <isidentical@gmail.com>
Co-authored-by: Pablo Galindo <Pablogsal@gmail.com>

Fidget-Spinner commented on Include/cpython/code.h:
```diff
@@ -165,6 +172,12 @@ PyCode_NewEmpty(const char *filename, const char *funcname, int firstlineno);
+/* Return the starting source code column offset from a bytecode index. */
+PyAPI_FUNC(int) PyCode_Addr2Offset(PyCodeObject*, int);
+/* Return the ending source code column offset from a bytecode index. */
+PyAPI_FUNC(int) PyCode_Addr2EndOffset(PyCodeObject*, int);
```
One question, doesn't this expose these functions as part of the public C API? Is that intentional?

encukou replied:
It doesn't: `Include/cpython/*` is ~~internal~~ *not limited* API.

Fidget-Spinner replied:
Ohh right, I forgot that for ``code.h`` in ``Python.h``, it's the one in ``Include/*``. Sorry and thanks for clarification Petr.

pablogsal replied:
@encukou  Is internal API but unless I am mistaken these are public functions. For instance, this file contains `PyCode_New` which is definitely public

pablogsal replied:
But in any case, yes, there is a new C-API function here (see the pep)

terryjreedy commented on Lib/test/test_compile.py:
```diff
@@ -985,6 +986,118 @@ def if_else_break():
+
+        return code, ast_tree
+
+    def assertOpCodeSourcePositionIs(self, code, opcode,
```
Here and below...
```suggestion
    def assertOpcodeSourcePositionIs(self, code, opcode,
```
is more readable to me.

terryjreedy commented on Lib/test/test_compile.py:
```diff
@@ -985,6 +986,118 @@ def if_else_break():
+        snippet += "while x != 0:\n"
+        snippet += "  x -= 1\n"
+        snippet += "while x != 0:\n"
+        snippet += "  x +=  1\n"
```
Copy 10000 lines just once.  To me, easier to read also.
```suggestion
        snippet += ("while x != 0:\n"
                    "  x -= 1\n"
                    "while x != 0:\n"
                    "  x +=  1\n"
                   )
```

terryjreedy reviewed (commented):
If I understand, the only officially visible change is the addition of code.co_positions, which returns an iterable of (start_line, end_line, start_column, end_column) tuples, where column numbers are 1-based.  Suggested title after bpo tag: "Add code.co_positions (PEP 657)" or "Add PEP-657 code.co_positions".

I presume a doc change will come.  I will review it for readability if pinged.

I read the tests and have two suggestions.
